download/buffer.go

```diff
@@ -85,88 +73,43 @@ func DownloadFileToBuffer(url string) (*bytes.Buffer, int64, error) {
-
-			success := false
-			for retryNum := 0; retryNum <= retries; retryNum++ {
-
-				if retryNum > 0 {
-					if verboseMode {
-						fmt.Printf("Retrying. Count: %d\n", retryNum)
-					}
-					sleepJitter := time.Duration(rand.Intn(retrySleepJitter))
-					sleepTime := time.Millisecond * (sleepJitter + retryDelayBaseline)
-
-					// Exponential backoff
-					// 2^retryNum * retryBackoffIncr (in milliseconds)
-					backoffFactor := math.Pow(retryBackoffFactor, float64(retryNum))
-					backoffDuration := time.Duration(math.Min(backoffFactor*retryBackoffIncr, retryMaxBackoffTime))
-					sleepTime += time.Millisecond * backoffDuration
-					time.Sleep(sleepTime)
-				}
-
-				transport := http.DefaultTransport.(*http.Transport).Clone()
-				transport.DisableKeepAlives = true
-				checkRedirectFunc := func(req *http.Request, via []*http.Request) error {
-					if verboseMode {
-						fmt.Printf("Received redirect '%d' to '%s'\n", req.Response.StatusCode, req.URL.String())
-					}
-					return nil
-				}
-				client := &http.Client{
-					Transport:     transport,
-					CheckRedirect: checkRedirectFunc,
-				}
-
-				req, err := http.NewRequest("GET", trueUrl, nil)
-				if err != nil {
-					// This needs to be a time.Duration to make everything happy
-					fmt.Printf("Error creating request: %v\n", err)
-					continue
-				}
-				req.Header.Set("Range", fmt.Sprintf("bytes=%d-%d", start, end))
-				if retryNum > 0 {
-					req.Header.Set("Retry-Count", fmt.Sprintf("%d", retryNum))
-				}
-
-				resp, err := client.Do(req)
-				if err != nil {
-					fmt.Printf("Error executing request: %v\n", err)
-					continue
-				}
-				defer resp.Body.Close()
-
-				n, err := io.ReadFull(resp.Body, data[start:end+1])
-				if err != nil && err != io.EOF {
-					fmt.Printf("Error reading response: %v\n", err)
-					continue
-				}
-				if n != int(end-start+1) {
-					fmt.Printf("Downloaded %d bytes instead of %d\n", n, end-start+1)
-					continue
-				}
-				success = true
-				break // if the download was successful, break out of the retry loop
-			}
-
-			if !success {
-				errc <- fmt.Errorf("failed to download after %d retries", retries)
-			}
-		}(start, end)
+		errGroup.Go(func() error {
+			return downloadChunk(ctx, start, end, data[start:end+1], trueURL)
+		})
 	}
 
-	wg.Wait()
-	close(errc) // close the error channel
-	for err := range errc {
-		if err != nil {
-			return nil, -1, err // return the first error we encounter
-		}
+	if err := errGroup.Wait(); err != nil {
+		return nil, -1, err // return the first error we encounter
 	}
+
 	elapsed := time.Since(startTime).Seconds()
 	througput := humanize.Bytes(uint64(float64(fileSize) / elapsed))
 	fmt.Printf("Downloaded %s bytes in %.3fs (%s/s)\n", humanize.Bytes(uint64(fileSize)), elapsed, througput)
 
 	buffer := bytes.NewBuffer(data)
 	return buffer, fileSize, nil
 }
+
+func downloadChunk(ctx context.Context, start, end int64, dataSlice []byte, trueURL string) error {
+	client := newClient()
```

Comment:
recreating the client for each chunk always bothered me but we can handle this once we get to downloading multiple files